pkg/controller/node/status.go

```diff
@@ -60,24 +89,91 @@ func calculateStatus(cconfig *v1.ControllerConfig, pool *mcfgv1.MachineConfigPoo
 	}
 	machineCount := int32(len(nodes))
 
-	updatedMachines := getUpdatedMachines(pool, nodes)
+	var degradedMachines, readyMachines, updatedMachines, unavailableMachines, updatingMachines []*corev1.Node
+	// if we represent updating properly here, we will also represent updating properly in the CO
+	// so this solves the cordoning RFE and the upgradeable RFE
+	// updating == updatePrepared, updateExecuted, updatedComplete
+	// updated == nodeResumed, updated
+	// ready == nodeResumed, updated
+	// unavailable == draining, cordoned
+	// degraded == if the condition.Reason == error
+	// this ensures that a MCP only enters Upgradeable==False if the node actually needs to upgrade to the new MC
+	for _, state := range mcs {
+		var ourNode *corev1.Node
+		for _, n := range nodes {
+			if state.Name == n.Name {
+				ourNode = n
+				break
+			}
+		}
+		if ourNode == nil {
+			klog.Errorf("Could not find specified node %s", state.Name)
+		}
+		if len(state.Status.Conditions) == 0 {
+			// not ready yet
+			break
+		}
+		for _, cond := range state.Status.Conditions {
+			if cond.Reason == "Error" {
+				degradedMachines = append(degradedMachines, ourNode)
+				continue
+			}
+			if cond.Status == metav1.ConditionUnknown {
+				switch v1alpha1.StateProgress(cond.Type) {
+				case mcfgalphav1.MachineConfigNodeUpdatePrepared:
+					updatingMachines = append(updatedMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateExecuted:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdatePostActionComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeResumed:
+					updatedMachines = append(updatingMachines, ourNode)
+					readyMachines = append(readyMachines, ourNode)
```

Comment:
To clarify, none of these states change from existing behaviour right?

Just thinking through the `readyMachines` state, for example, which it looks like we have to be updated to be considered Ready. And if all the nodes are accounted for, we are happy (and skip the getReadyMachines call which now only happens if the nodes count for each status don't add up to the total)

This means that in scenarios like:
```
		if cond.Type == corev1.NodeReady && cond.Status != corev1.ConditionTrue {
			return fmt.Errorf("node %s is reporting NotReady=%v", node.Name, cond.Status)
		}
		if cond.Type == corev1.NodeDiskPressure && cond.Status != corev1.ConditionFalse {
			return fmt.Errorf("node %s is reporting OutOfDisk=%v", node.Name, cond.Status)
		}
		if cond.Type == corev1.NodeNetworkUnavailable && cond.Status != corev1.ConditionFalse {
			return fmt.Errorf("node %s is reporting NetworkUnavailable=%v", node.Name, cond.Status)
		}
```

Will be no longer accounted for if the nodes are otherwise in a "completed" state? As in previously, even if a node is updated, if it then reports say, disk pressure, the MCO will count it as unready again. Here, this would not happen always?

Let me know if that doesn't make sense. I'm trying to wrap my head around the subtle differences of state, and want to make sure we don't change any behaviour we don't intend to. And for differences we do want, we should explicitly call out.